### PR TITLE
netif: add functions to get and get by identifier 

### PR DESCRIPTION
**Pull request by miri64**
### Contribution description
This is a proof-of-concept implementation for a stack-independent interface identifiers.

### Testing procedure
None yet, but will add some if we agree on the concept.

### Issues/PRs references
A proposed solution for #12680.

### CODE REVIEW

**Comment by kaspar030** [comment]
This'll traverse the netif list on every call (in O(N) time). IMO, for something that might be called multiple times per packet, that's not acceptable. Think benchmarks having different results depending on the number of configured network devices.

Why not an extra field in netif_t for the numerical id (for netif_get_id()), and a mapping array (for netif_by_id())?

**Comment by miri64** [comment]
> This'll traverse the netif list on every call (in O(N) time). IMO, for something that might be called multiple times per packet, that's not acceptable. Think benchmarks having different results depending on the number of configured network devices.

I don't think it will be called multiple times per packet. I expect `netif_get_id()` to be called at most once per end-point initialization and `netif_get_by_pid()` once in the translation.

> Why not an extra field in netif_t for the numerical id (for netif_get_id()), and a mapping array (for netif_by_id())?

I try to prevent topping the number of network interfaces at compile time. We currently actively go away from that.

**Comment by miri64** [comment]
Or we do it like I proposed originally (see latest fixup). This would be O(1) but also relies on some assumptions (all `netif_t` instances can be distinguished by the same 16-bit address suffix and `netif_get_id()` is only called for existing network interfaces.

**Comment by miri64** [comment]
Ping?

**Comment by miri64** [comment]
Ping @leandrolanzieri @jia200x?

**Comment by leandrolanzieri** [comment]
> I don't think it will be called multiple times per packet. I expect `netif_get_id()` to be called at most once per end-point initialization and `netif_get_by_pid()` once in the translation.

If this is the case, it seems to me that the cleaner approach is the one proposed in 4ef741febcf73d667bbe3df158dc36aa7401f269. 

> I try to prevent topping the number of network interfaces at compile time. We currently actively go away from that.

+1!

**Comment by miri64** [comment]
> > I don't think it will be called multiple times per packet. I expect `netif_get_id()` to be called at most once per end-point initialization and `netif_get_by_pid()` once in the translation.
> 
> If this is the case, it seems to me that the cleaner approach is the one proposed in [4ef741f](https://github.com/RIOT-OS/RIOT/commit/4ef741febcf73d667bbe3df158dc36aa7401f269).

Okay, I reverted c3cab1c803f70d9fa1daf14c43a1f6fed6fdcf0c

File: sys/net/netif/netif.c [hunk 1]
```diff
@@ -57,4 +68,16 @@ netif_t *netif_get_by_name(const char *name)
 
     return NULL;
 }
+
+int16_t netif_get_by_id(int16_t netif)
```
Comment (leandrolanzieri) [inline sys/net/netif/netif.c]:
```suggestion
netif_t *netif_get_by_id(int16_t id)
```

File: sys/net/netif/netif.c [hunk 2]
```diff
@@ -57,4 +68,16 @@ netif_t *netif_get_by_name(const char *name)
...
+{
+    list_node_t *node = netif_list.next;
+    for (int16_t i = 0; node; i++, node = node->next) {
+        if (i == id) {
+            return (netif_t *)node;
+        }
+    }
+    return -1;
```
Comment (leandrolanzieri) [inline sys/net/netif/netif.c]:
```suggestion
    return NULL;
```

File: sys/net/netif/netif.c (hunk 1 above)
Comment (miri64) [reply sys/net/netif/netif.c]: Addressed.

File: sys/net/netif/netif.c (hunk 2 above)
Comment (miri64) [reply sys/net/netif/netif.c]: Addressed

**Comment by benpicco** [comment]
IMHO you can squash.

**Comment by miri64** [comment]
I squashed the changes so far, but, and I know this PR was already ack'd, I got an idea, how to keep the id thing downwards compatible, when used with `sock`: Make the definitions introduced here `weak` and override them in GNRC with the PID equivalent (`netif->pid` for `netif_get_id()` and `gnrc_netif_get_by_id()` for `netif_get_by_id()`. Let me know what you think.

**Comment by miri64** [comment]
I tested if this works with the following patch

```diff
diff --git a/examples/gnrc_networking/main.c b/examples/gnrc_networking/main.c
index 6301f4291..3d5cc9c63 100644
--- a/examples/gnrc_networking/main.c
+++ b/examples/gnrc_networking/main.c
@@ -22,6 +22,7 @@
 
 #include "shell.h"
 #include "msg.h"
+#include "net/netif.h"
 
 #define MAIN_QUEUE_SIZE     (8)
 static msg_t _main_msg_queue[MAIN_QUEUE_SIZE];
@@ -35,11 +36,15 @@ static const shell_command_t shell_commands[] = {
 
 int main(void)
 {
+    netif_t *netif = NULL;
     /* we need a message queue for the thread running the shell in order to
      * receive potentially fast incoming networking packets */
     msg_init_queue(_main_msg_queue, MAIN_QUEUE_SIZE);
     puts("RIOT network stack example application");
 
+    while ((netif = netif_iter(netif))) {
+        printf("netif %i\n", netif_get_id(netif));
+    }
     /* start shell */
     puts("All up, running the shell now");
     char line_buf[SHELL_DEFAULT_BUFSIZE];
```

**Comment by benpicco** [comment]
But now the ID is not predictable anymore.
I liked this PR because it enabled pid-independent IDs for interfaces.

Now it will again change based on what modules are used, so it can not be used to attach information to IDs on the board level.

**Comment by miri64** [comment]
> But now the ID is not predictable anymore.
> I liked this PR because it enabled pid-independent IDs for interfaces.
> 
> Now it will again change based on what modules are used, so it can not be used to attach information to IDs on the board level.

With the current approach however we can slowly phase over to that instead of having to change over in a huge, hardly testable, push.

**Comment by miri64** [comment]
IMHO: since `sock` is a very stable API, that is expected to be used outside of the main repo, this is the more desirable option.

**Comment by benpicco** [comment]
But changing the behavior of `netif_get_id()` depending on what stack is used sounds like a bad idea.

How about leaving `netif_get_id()` as is and instead provide a stack-dependent wrapper function?

```C
static inline int16_t gnrc_netif_get_id(const gnrc_netif_t *netif)
{
    return netif->pid;
}

static inline int16_t lwip_netif_get_id(const netif_t *netif)
{
    return netif_get_id(netif);
}
```

**Comment by miri64** [comment]
> But changing the behavior of `netif_get_id()` depending on what stack is used sounds like a bad idea.

The fallback (or previous general approach) is also stack dependent, because it is dependent in which order the interfaces are initialized (which at the moment the stacks do).

> How about leaving `netif_get_id()` as is and instead provide a stack-dependent wrapper function?

That would defeat the purpose of a generalized API used with `sock`. Because then we would need do something like this in the transition period:

```C
#if IS_USED(MODULE_GNRC)
    int16_t id = gnrc_netif_get_id(netif);
#elif IS_USED(MODULE_LWIP)
    int16_t id = lwip_netif_get_id(netif);
#endif
```

Then I rather don't introduce this function at all until we are transition and splatter this all over the code base ;-P:

```C
#if IS_USED(MODULE_GNRC)
    int16_t id = ((gnrc_netif_t *)netif)->pid;
#elif IS_USED(MODULE_LWIP)
    int16_t id = ((struct netif *)netif)->num;
#endif
```

**Comment by benpicco** [comment]
Hm and what I was trying to do wouldn't work anyway since there is currently no way to get the `netif_t *` corresponding to a `netdev_t *` :confused:

**Comment by miri64** [comment]
Yes, we are still a bit away from compile-time-unique identifiers. But this was not the point of this PR anyway (this as the OP says is more a user-level API), but with this we might become closer.

**Comment by miri64** [comment]
Squashed

**Comment by aabadie** [comment]
@benpicco @miri64 is this one still mergeable ?

**Comment by miri64** [comment]
IMHO, yes.

**Comment by miri64** [comment]
I give it another rebuild though.

**Pull request merged**